src/oas/resolver.ts

```diff
@@ -47,27 +57,46 @@ export const bundleResolveRef: RefResolver = function (target) {
   return syncReferenceObject(target, this.references);
 };
 
-function getResolved(references: References, $ref: string) {
-  const seen = new Set();
-  let value = $ref;
+/**
+ * Find the nearest resolved reference URI, starting with `$ref` and stepping
+ * through all known references.
+ * @param references all known references
+ * @param startingRefUri the starting reference URI
+ * @returns the first resolved reference URI, or unknown reference URI
+ */
+function getResolved(references: References, startingRefUri: string): string {
+  const seen = new Set(); // for cycle detection
+
+  let refUri: string = startingRefUri;
+  while (refUri in references) {
+    // don't follow cycles; bail out immediately
+    if (seen.has(refUri)) {
+      return refUri;
+    }
+
+    seen.add(refUri);
 
-  while (value in references) {
-    if (seen.has(value)) return value;
-    seen.add(value);
+    // follow the reference one step
+    const referenced = references[refUri];
+    refUri = referenced.value;
 
-    const referenced = references[value];
-    ({ value } = referenced);
+    // return the first resolved reference we encounter
     if (referenced.resolved) {
-      break;
+      return refUri;
     }
   }
 
-  return value;
+  // we've reached a completely unresolved *and* unknown reference URI
+  return refUri;
```

Comment:
Added lots of comments, 'cause I didn't understand what this function was doing or how.